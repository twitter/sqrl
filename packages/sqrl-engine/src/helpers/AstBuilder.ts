/**
 * Copyright 2019 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import {
  ConstantAst,
  SlotAst,
  Ast,
  CallAst,
  IfAst,
  ListAst,
  FeatureAst
} from "../api/ast";
import { ConstantSlot } from "../api/parse";
import { invariant } from "sqrl-common";

/**
 * Helper class to construct new Ast objects
 */
export class AstBuilder {
  static constant(value: any): ConstantAst {
    return {
      type: "constant",
      value
    };
  }

  static slot(value: ConstantSlot): SlotAst {
    return {
      type: "slot",
      slotName: value._wrapped.name
    };
  }

  static call(functionName: string, args: Ast[]): CallAst {
    return {
      type: "call",
      func: functionName,
      args
    };
  }

  static props(obj: { [key: string]: Ast }): CallAst {
    const args: Ast[] = [];
    Object.entries(obj).forEach(([key, ast]) => {
      args.push(AstBuilder.constant(key), ast);
    });
    return AstBuilder.call("createMap", args);
  }

  static branch(condition: Ast, trueBranch: Ast, falseBranch: Ast): IfAst {
    return {
      type: "if",
      condition,
      trueBranch,
      falseBranch
    };
  }

  static feature(name: string): FeatureAst {
    return { type: "feature", value: name };
  }

  static and(asts: Ast[]): Ast {
    invariant(
      asts.length > 0,
      "SQRL AstBuilder requires at-least one value for an `and`"
    );
    let result: Ast = AstBuilder.constant(true);
    for (let idx = asts.length - 1; idx >= 0; idx--) {
      const ast: Ast = asts[idx];
      if (ast.type === "constant") {
        if (ast.value) {
          continue;
        } else {
          return AstBuilder.constant(false);
        }
      } else {
        result =
          result.type === "constant" && result.value === true
            ? ast
            : {
                type: "boolean_expr",
                operator: "and",
                left: ast,
                right: result
              };
      }
    }
    return result;
  }

  static or(asts: Ast[]): Ast {
    invariant(
      asts.length > 0,
      "SQRL AstBuilder requires at-least one value for an `or`"
    );
    let result: Ast = AstBuilder.constant(false);
    for (let idx = asts.length - 1; idx >= 0; idx--) {
      const ast: Ast = asts[idx];
      if (ast.type === "constant") {
        if (!ast.value) {
          continue;
        } else {
          return AstBuilder.constant(true);
        }
      } else {
        result =
          result.type === "constant" && result.value === false
            ? ast
            : {
                type: "boolean_expr",
                operator: "or",
                left: ast,
                right: result
              };
      }
    }
    return result;
  }

  static list(asts: Ast[]): ListAst {
    return {
      type: "list",
      exprs: asts
    };
  }
}
