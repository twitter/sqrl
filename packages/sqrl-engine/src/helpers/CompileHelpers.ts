/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { LocalFilesystem, Filesystem } from "../api/filesystem";
import { SqrlParserSourceOptions } from "../compile/SqrlParserState";
import { basename, dirname } from "path";
import { parseSqrl } from "../parser/SqrlParse";
import { StatementAst } from "../ast/Ast";
import invariant from "../jslib/invariant";

export function statementsFromString(
  source: string,
  options: {
    customFunctions?: Set<string>;
  } = {}
): StatementAst[] {
  return parseSqrl(source, {
    customFunctions: options.customFunctions
  }).statements;
}

export function sourceOptionsFromFilesystem(
  filesystem: Filesystem,
  mainFilename: string,
  options: {
    customFunctions?: Set<string>;
  } = {}
) {
  const sourceBuffer = filesystem.tryRead(mainFilename);
  invariant(sourceBuffer, "Could not read main file source: %s", mainFilename);
  const source = sourceBuffer.toString("utf-8");
  const statements = parseSqrl(source, {
    filename: mainFilename,
    customFunctions: options.customFunctions
  }).statements;
  return {
    statements,
    source,
    filesystem
  };
}
export function sourceOptionsFromPath(
  path: string,
  options: {
    customFunctions?: Set<string>;
  } = {}
): SqrlParserSourceOptions {
  return sourceOptionsFromFilesystem(
    new LocalFilesystem(dirname(path)),
    basename(path),
    options
  );
}
