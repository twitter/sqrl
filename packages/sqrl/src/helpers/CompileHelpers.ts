/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { LocalFilesystem, Filesystem } from "../api/filesystem";
import {
  SqrlParserSourceOptions,
  SqrlParserState
} from "../compile/SqrlParserState";
import { basename, dirname } from "path";
import { parseSqrl } from "../parser/SqrlParse";
import FunctionRegistry from "../function/FunctionRegistry";
import SqrlExecutable from "../execute/SqrlExecutable";
import { compileParserStateAst } from "../compile/SqrlCompile";
import { SqrlCompiledOutput } from "../compile/SqrlCompiledOutput";
import { LabelerSpec } from "../execute/LabelerSpec";
import { JsExecutionContext } from "../execute/JsExecutionContext";
import { StatementAst } from "../ast/Ast";
import invariant from "../jslib/invariant";
import { Context } from "../api/ctx";
import { FeatureMap } from "../api/execute";

export function statementsFromString(source: string): StatementAst[] {
  return parseSqrl(source).statements;
}

export function sourceOptionsFromFilesystem(
  filesystem: Filesystem,
  mainFilename: string
) {
  const sourceBuffer = filesystem.tryRead(mainFilename);
  invariant(sourceBuffer, "Could not read main file source: %s", mainFilename);
  const source = sourceBuffer.toString("utf-8");
  const statements = parseSqrl(source, {
    filename: mainFilename
  }).statements;
  return {
    statements,
    source,
    filesystem
  };
}
export function sourceOptionsFromPath(path: string): SqrlParserSourceOptions {
  return sourceOptionsFromFilesystem(
    new LocalFilesystem(dirname(path)),
    basename(path)
  );
}

export async function executableFromString(
  source: string,
  options: {
    trace?: Context;
    functionRegistry: FunctionRegistry;
    setInputs?: FeatureMap;
    librarySource?: string;
  }
): Promise<SqrlExecutable> {
  let libraryStatements: StatementAst[] = [];
  if (options.librarySource) {
    libraryStatements = [...statementsFromString(options.librarySource)];
  }

  const statements: StatementAst[] = [
    ...libraryStatements,
    ...statementsFromString(source)
  ];
  const parserState = new SqrlParserState({
    statements,
    functionRegistry: options.functionRegistry,
    setInputs: options.setInputs || {}
  });
  compileParserStateAst(parserState);
  const compiledOutput = new SqrlCompiledOutput(parserState);
  const spec = await compiledOutput.buildLabelerSpec(options.trace);
  return executableFromSpec(options.functionRegistry, spec);
}

export function executableFromSpec(
  functionRegistry: FunctionRegistry,
  spec: LabelerSpec
) {
  const context = new JsExecutionContext(functionRegistry);
  return new SqrlExecutable(context, spec);
}
