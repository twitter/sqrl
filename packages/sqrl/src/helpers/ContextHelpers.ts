/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { SimpleDatabaseSet } from "../platform/DatabaseSet";
import { getGlobalLogger } from "../api/log";
import { SimpleContext } from "../platform/Trace";

// With many customers we can create new dataset id's, for now scope it down to 1
export const DEFAULT_DATASET_ID = "1";

export function createDefaultContext() {
  return new SimpleContext(
    new SimpleDatabaseSet(DEFAULT_DATASET_ID),
    getGlobalLogger()
  );
}
