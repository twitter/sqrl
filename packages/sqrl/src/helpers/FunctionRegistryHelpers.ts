import { SqrlFunctionRegistry } from "../function/FunctionRegistry";
import {
  registerAllFunctions,
  FunctionServices
} from "../function/registerAllFunctions";

export function buildFunctionRegistryForServices(services?: FunctionServices) {
  const registry = new SqrlFunctionRegistry();
  registerAllFunctions(registry, services);
  return registry;
}
