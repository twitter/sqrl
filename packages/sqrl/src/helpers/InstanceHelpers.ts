/**
 * Copyright 2019 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { SqrlInstance } from "../function/Instance";
import { registerAllFunctions } from "../function/registerAllFunctions";
import { FunctionServices } from "../api/execute";

export function buildSqrlInstanceForServices(services: FunctionServices) {
  const instance = new SqrlInstance();
  registerAllFunctions(instance);
  return instance;
}
