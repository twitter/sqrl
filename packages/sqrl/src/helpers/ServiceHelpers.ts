/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import {
  FunctionServices,
  KafkaService
} from "../function/registerAllFunctions";
import { AssertService } from "../function/AssertFunctions";
import { BlockService } from "../function/BlockFunctions";
import { CountService } from "../function/CountFunctions";
import { CountUniqueService } from "../function/CountUniqueFunctions";
import { LabelService } from "../function/LabelFunctions";
import { PatternService } from "../function/PatternFunctions";
import { UniqueIdService } from "../function/NodeFunctions";
import { RateLimitService } from "../function/RateLimitFunctions";
import { ObjectService } from "../function/SaveFunctions";
import { InProcessPatternService } from "../services/InProcessPatternService";
import { SmyteRateLimit } from "../services/SmyteRateLimit";
import { RedisInterface, RedisService } from "../services/RedisService";
import { MockRedisService } from "../mocks/MockRedisService";
import { RedisCountService } from "../services/RedisCountService";
import { RedisApproxCountUniqueService } from "../services/RedisApproxCountUnique";
import { RedisLabelService } from "../services/RedisLabelService";
import { RedisUniqueIdService } from "../services/RedisUniqueId";
import { RedisRateLimit } from "../services/RedisRateLimit";

interface Closeable {
  close(): void;
}

class SqrlStandardServices implements FunctionServices {
  assert?: AssertService;
  block?: BlockService;
  count?: CountService;
  countUnique?: CountUniqueService;
  label?: LabelService;
  pattern?: PatternService;
  uniqueId?: UniqueIdService;
  rateLimit?: RateLimitService;
  saveFeatures?: KafkaService;
  object?: ObjectService;

  private shutdown: Closeable[] = [];
  constructor(props: {
    ratelimitAddress?: string;
    redisAddress?: string;
    inMemory?: boolean;
  }) {
    const { ratelimitAddress, redisAddress, inMemory } = props;
    this.pattern = new InProcessPatternService();

    let redisService: RedisInterface = null;
    if (redisAddress) {
      const redis = new RedisService(redisAddress);
      redisService = redis;
      this.shutdown.push(redis);
    } else if (inMemory) {
      redisService = new MockRedisService();
    }

    if (ratelimitAddress) {
      const rateLimit = new SmyteRateLimit(ratelimitAddress);
      this.rateLimit = rateLimit;
      this.shutdown.push(rateLimit);
    }

    if (redisService) {
      this.count = new RedisCountService(redisService, "count~");
      this.countUnique = new RedisApproxCountUniqueService(
        redisService,
        "countUnique~"
      );
      this.label = new RedisLabelService(redisService, "label~");
      this.uniqueId = new RedisUniqueIdService(redisService, "id~");

      if (!this.rateLimit) {
        this.rateLimit = new RedisRateLimit(redisService, "ratelimit~");
      }
    }
  }

  close() {
    this.shutdown.forEach(svc => svc.close());
  }
}

export function buildServicesFromAddresses(props: {
  ratelimitAddress?: string;
  redisAddress?: string;
  inMemory?: boolean;
}) {
  return new SqrlStandardServices(props);
}
