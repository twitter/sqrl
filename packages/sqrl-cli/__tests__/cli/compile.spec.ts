/**
 * Copyright 2019 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */

import { runCompile } from "../helpers/runCompile";

test("repl works", async () => {
  const output = await runCompile();
  expect(JSON.parse(output)).toMatchSnapshot();
});
