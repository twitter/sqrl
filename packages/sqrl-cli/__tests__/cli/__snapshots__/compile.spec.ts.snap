// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repl works 1`] = `
Object {
  "slotExprs": Array [
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "exprs": Array [],
      "func": "now",
      "type": "call",
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "constant",
      "value": true,
    },
    Object {
      "type": "input",
    },
    Object {
      "exprs": Array [
        Object {
          "exprs": Array [
            Object {
              "type": "constant",
              "value": "Hello, ",
            },
            Object {
              "slot": 10,
              "type": "value",
            },
            Object {
              "type": "constant",
              "value": "!",
            },
          ],
          "func": "concat",
          "type": "call",
        },
      ],
      "load": Array [
        10,
      ],
      "type": "load",
    },
    Object {
      "type": "constant",
      "value": true,
    },
  ],
  "slotNames": Array [
    "SqrlMutate",
    "SqrlIsClassify",
    "SqrlClock",
    "SqrlExecutionComplete",
    "SqrlAssertionStatements",
    "SqrlCountStatements",
    "SqrlCountUniqueStatements",
    "SqrlLabelStatements",
    "SqrlLogStatements",
    "SqrlBlockStatements",
    "Name",
    "Text",
    "bool(:)",
  ],
}
`;
